stackslib/src/net/download/nakamoto.rs

```diff
@@ -0,0 +1,4283 @@
+    pub(crate) fn add_downloaders(
+        &mut self,
+        iter: impl IntoIterator<Item = (NeighborAddress, NakamotoTenureDownloader)>,
+    ) {
+        for (naddr, downloader) in iter {
+            if self.has_downloader(&naddr) {
+                test_debug!("Already have downloader for {}", &naddr);
+                continue;
+            }
+            self.add_downloader(naddr, downloader);
+        }
+    }
+
+    /// Count up the number of in-flight messages, based on the states of each instantiated
+    /// downloader.
+    pub fn inflight(&self) -> usize {
+        let mut cnt = 0;
+        for downloader_opt in self.downloaders.iter() {
+            let Some(downloader) = downloader_opt else {
+                continue;
+            };
+            if downloader.idle {
+                continue;
+            }
+            if downloader.is_waiting() {
+                continue;
+            }
+            if downloader.is_done() {
+                continue;
+            }
+            cnt += 1;
+        }
+        cnt
+    }
+
+    /// Determine whether or not there exists a downloader for the given tenure, identified by its
+    /// consensus hash.
+    pub fn is_tenure_inflight(&self, ch: &ConsensusHash) -> bool {
+        for downloader_opt in self.downloaders.iter() {
+            let Some(downloader) = downloader_opt else {
+                continue;
+            };
+            if &downloader.tenure_id_consensus_hash == ch {
+                return true;
+            }
+        }
+        false
+    }
+
+    /// Determine if this downloader set is empty -- i.e. there's no in-flight requests.
+    pub fn is_empty(&self) -> bool {
+        self.inflight() == 0
+    }
+
+    /// Try to resume processing a download state machine with a given peer.  Since a peer is
+    /// detached from the machine after a single RPC call, this call is needed to re-attach it to a
+    /// (potentially different, unblocked) machine for the next RPC call to this peer.
+    ///
+    /// Returns true if the peer gets scheduled.
+    /// Returns false if not.
+    pub fn try_resume_peer(&mut self, naddr: NeighborAddress) -> bool {
+        if let Some(idx) = self.peers.get(&naddr) {
+            let Some(Some(_downloader)) = self.downloaders.get(*idx) else {
+                return false;
+            };
+
+            test_debug!(
+                "Peer {} already bound to downloader for {}",
+                &naddr,
+                &_downloader.tenure_id_consensus_hash
+            );
+            return true;
+        }
+        for (i, downloader_opt) in self.downloaders.iter_mut().enumerate() {
+            let Some(downloader) = downloader_opt else {
+                continue;
+            };
+            if !downloader.idle {
+                continue;
+            }
+            if downloader.is_waiting() {
+                continue;
+            }
+            if downloader.naddr != naddr {
+                continue;
+            }
```

Comment:
Why does downloader have its own `naddr` field, and where is it being set?

Reply:
It's being used to "resume" downloads when there are fewer peers than in-flight downloads.  Consider the simple case where there's only one peer, but six downloaders (this is what happens in the unit test):

* Peer A is assigned to D1
* Peer A fetches D1's tenure-start block, transitioning D1 to `WaitForTenureEndBlock`
* D1 is now waiting and idle, so peer A is detached from it
* Downloader D2 is instantiated by `NakamotoTenureDownloaderSet` to fetch the child tenure of D1
* Peer A is reassigned to D2
* Peer A fetches D2's tenure-start block, which puts D2 into `WaitForTenureEndBlock`
* D2 is now waiting and idle, so peer A is detached from it
* The `NakamotoTenureDownloaderSet` machine notices that D2 has the tenure-end block for D1, and passes it to D1, which transitions D1 to `GetTenureBlocks`
* Peer A is re-assigned to D1
* Peer A fetches the tenure blocks for D1, which transitions D1 to `Done`
* Peer A is detached from D1, and D1 is shut down
* Downloader D3 is instantiated by `NakamotoTenureDownloaderSet` to fetch the child tenure of D2

... and so on and so forth, until all of the reward cycle's tenures are downloaded by up to six `NakamotoTenureDownloader` machines driven by `NakamotoTenureDownloaderSet`.  As you can see, the job of the `NakamotoTenureDownloaderSet` is to assign downloaders to peers in order to both issue the HTTP requests for the downloaders and transition them to their next states.  The act of detaching and re-attaching peers to downloaders is what enables the system to make progress even when some downloaders are blocked on data that will eventually become available from other downloaders (e.g. the dependence of D2 on D1).

You'd want the `NakamotoTenureDownloader` to "stick" to a peer via this check because you've already verified that that peer can serve that tenure.  It's not guaranteed that each peer can serve each tenure; some peers may be booting up, for example.  So, the `NakamotoTenureDownloaderSet` will only ever assign a `NakamotoTenureDownloader` machine to a single peer over its lifetime.  If that machine fails to run to completion, the top-level `NakamotoDownloaderStateMachine` will request the `PeerNetwork` to disconnect from the offending peer, causing a new `NakamotoTenureDownloader` to be instantiated and assigned to a different peer on the next pass.